### ISSUE TITLE
Rotating verdaccio logs

### ISSUE THREAD

**Issue by UnitedMarsupials** [verdaccio/verdaccio#420]
This is the same question I asked for rlidwka/sinopia#472 -- I'm configuring Verdaccio to log into a file, which will need to be rotated.

Is there some way to signal to the program to reopen the log, or do I have to fully restart it? Thank you!

**Comment by juanpicado** [comment]
I'm not 100% sure what the answer here. The PR #408 is working on a new log system based on `pino`.  Perhaps @jsumners has a better answer.

**Comment by UnitedMarsupials** [comment]
Wow... Am I really the first person to ask this question?..

**Comment by jsumners** [comment]
If #408 is merged in, then you will rotate your logs via the normal system mechanisms -- http://getpino.io/#/docs/howtos?id=rotate

**Comment by UnitedMarsupials** [comment]
The document you linked to prescribes logging to `stdout` and using the `copytruncate` verb with logrotate.

I would strongly advise _against_ incorporating any such mechanisms.

**Comment by jsumners** [comment]
Uh, why? It's worked fine for decades.

**Comment by UnitedMarsupials** [comment]
Because, instead of moving the it aside, `copytruncate` creates a _copy_ of the log file and then proceeds to truncate the original so the application’s next message will be at the beginning of the file.

This is wasteful - creating a copy is immensely more expensive than simply issuing a `rename()`. 

It is also likely to cause [data loss](https://incoherency.co.uk/blog/stories/logrotate-copytruncate-race-condition.html) if the app decides to log something after the copying is over but before the truncation is done. As with all race conditions, such loss will be hard to diagnose and reproduce. 

Copytruncate is a crutch created for people dealing with applications that don’t implement log-rotation properly. Verdaccio does not need to join in this infamy.

A proper server would log through syslog or reopen its logs upon receiving a signal, such as `USR1`.

I urge you to read up on this before making any changes to an application’s logging mechanism.

**Comment by jsumners** [comment]
_Applications_ should not implement log rotation. They should write logs and the system should rotate them however it sees fit. If you want to write to syslog with #408 merged, just `node server.js | pino-syslog > /dev/log`.

https://www.npmjs.com/package/pino-syslog

**Comment by UnitedMarsupials** [comment]
> Applications should not implement log rotation.

This may seem self evident to you, but it is a completely bogus statement.

But I’ll be happy, if all applications used syslog.

Just not in the manner you suggested because that would necessitate an extra context-switch for every message logged.

Anyway, I hope, I’ve convinced you, anything relying on `copytruncate` is incorrect. If pino can do syslog directly (without piping to another process), that’s fine too.

**Comment by UnitedMarsupials** [comment]
>>  _Applications_ should not implement log rotation.

> This may seem self evident to you, but it is a completely bogus statement.

Sorry, let me rephrase it. As stated, I agree with the notion in itself -- although developers of [log4j](http://2min2code.com/articles/log4j_intro/rolling_file_on_size_prop) might disagree, applications should not indeed do their own log-rotation.

But -- and this is why I called the statement "bogus" -- that's not at all, what I'm advocating here. I do not ask for Verdaccio -- nor _any_ application -- to rotate the logs. You are correct, that this is best left for the OS tools such as [logrotate](/logrotate/logrotate) or [newsyslog](https://www.newsyslog.org/newsyslog.html).

However, for any such OS tool do be able to do its job, a modicum of cooperation from the application is necessary. The normal way for most applications to do this is to implement a signal-handler for a documented signal (such as `HUP` or `USR1`). Upon receiving such signal, the application should make a note to reopen the currently opened log(s) before outputing the next log-message.

Without such cooperation, the log-rotating software can only do some variation of `copytruncate`, which is both wasteful and dangerous.

Do a `grep kill /etc/logrotate.d/*` on your nearest Linux machine. Verdaccio ought to have a similar entry.

**Comment by jsumners** [comment]
>  > Applications should not implement log rotation.
>
> This may seem self evident to you, but it is a completely bogus statement. 

Not, it's not. That is the whole point of the logrotate utility. Applications should focus on their actual purpose. Rotating logs is outside of the scope of all applications _except_ those that are designed to manage logs.

In particular for applications written in Javascript on top of Node.js: these are single threaded applications (basically). They should definitely be expending their cycles on their actual work in order to properly service the requests sent to them. If your system can't handle process 1 writing logs to process 2, then you have bigger issues than using a pipe to deal with logs via `stdout`.

**Comment by UnitedMarsupials** [comment]
>>  _Applications_ should not implement log rotation.

> This may seem self evident to you, but it is a completely bogus statement.

Sorry, let me rephrase it. As stated, I agree with the notion in itself -- although developers of [log4j](http://2min2code.com/articles/log4j_intro/rolling_file_on_size_prop) might disagree, applications should not indeed do their own log-rotation.

But -- and this is why I called the statement "bogus" -- that's not at all, what I'm advocating here. I do not ask for Verdaccio -- nor _any_ application -- to rotate the logs. You are correct, that this is best left for the OS tools such as [logrotate](/logrotate/logrotate) or [newsyslog](https://www.newsyslog.org/newsyslog.html).

However, for any such OS tool do be able to do its job, a modicum of cooperation from the application is necessary. The normal way for most applications to do this is to implement a signal-handler for a documented signal (such as `HUP` or `USR1`). Upon receiving such signal, the application should make a note to reopen the currently opened log(s) before outputing the next log-message.

Without such cooperation, the log-rotating software can only do some variation of `copytruncate`, which is both wasteful and dangerous.

Do a `grep kill /etc/logrotate.d/*` on your nearest Linux machine. Verdaccio ought to have a similar entry.

**Comment by UnitedMarsupials** [comment]
Looking inside Verdaccio source, I see that it currently uses Buyan logger -- which does support reopening the log-files upon receiving a signal.

FWIW, I filed trentm/node-bunyan#560 asking for advice on how to add such handling to `logger.js` -- should not be a very large change...

**Comment by UnitedMarsupials** [comment]
So, the answer my own question for the record is:

- be sure to use Verdaccio above 2.6.6
- signal the verdaccio process with `SIGUSR2` to cause it to reopen the log-file

For example, when using logrotate, have an the `postrotate` entry like this:
```
        postrotate
                exec killall -USR2 verdaccio
        endscript
```

**Comment by juanpicado** [comment]
released `v2.7.0`
